### ISSUE TITLE
Need smarter contextmenu items

### ISSUE THREAD

**Issue by telamonian** [phosphorjs/phosphor#354]
People working on the context menu of JupyterLab have run into various problems (see jupyterlab/jupyterlab#4529 for a good overview). After spending a while thinking about it, I believe that all of the issues can be resolved by making the contextmenu items in `@phorsphor/widgets` a little bit smarter/more aware of the circumstances under which they were invoked. Specifically, I think that two changes are required:

1. There needs to be a way to pass information contained in the node that the user actually right-clicked on to the command of a contextmenu item.

2. There needs to be a way to show/hide contextmenu items based on (again) information contained in the node that the user actually right-clicked on.

**Comment by telamonian** [comment]
I've already written a PR (#355) addressing the first point.

**Comment by blink1073** [comment]
For 2, you can use CSS dataset selectors in the registered context menu item.

**Comment by blink1073** [comment]
Hmm, as long as the context menu handler is scoped to an appropriate selector (not `body`), I think it is fair to give it a chance to show or not based on the current event target (performance-wise).

**Comment by sccolbert** [comment]
I agree with the thesis of this thread. We've known this for a while. So far nothing has hit me as the "right" way to provide that information. I agree with @blink1073 that CSS selectors *should* be enough to control visibility, provided you tag your node with enough information.

My current thinking is I'd like to just find a clean way to pass the DOM node to command which is executed, but there may be some unintended consequences to that, such as what to pass when the command is invoked directly?

**Comment by sccolbert** [comment]
At this point we are not getting rid of selectors, since that would be a breaking change.

**Comment by telamonian** [comment]
>but there may be some unintended consequences to that, such as what to pass when the command is invoked directly??

My thoughts exactly. When I set out to write #355 I did originally intended to just have it pass the target node to a command, but then I realized that would make it awkward to write commands that could both be used in contextmenu items and invoked directly. The design I settled on is intended to make invocation simple, since this way the names of the dataset properties you'd set when invoking a command via a context menu would be the same as the names of the args you'd pass when invoking the command directly.

Still, if you think it's better to pass the whole node, I guess it wouldn't be that big a deal? You would just need to test for/branch on the presence of an optional `node` arg (or static property, or whatever) within the body of relevant commands.

**Comment by hadim** [comment]
What is the status of this issue? I am looking for a way to select only certain file according to their extension ('.zip', 'tar,bz2', etc). See https://github.com/hadim/jupyter-archive/issues/19 for details.

In CSS you can add multiple selectors with a comma: const selectorOnlyArchive = '.jp-DirListing-item[title$=".zip"], .jp-DirListing-item[title$=".tgz"]';

But it does not seem to work in Jupyter. And actually the documentation is pretty clear about it: https://phosphorjs.github.io/phosphor/api/widgets/interfaces/contextmenu.iitemoptions.html

**The selector must not contain commas.**

So is there is a way with the current API to achieve this kind of selection only with the `selector` attribute?

**Comment by hadim** [comment]
Indeed this is what I ended up doing. See https://github.com/hadim/jupyter-archive/issues/19

**Comment by telamonian** [comment]
> provided you tag your node with enough information.

@hadim Following this piece of advice from @sccolbert was ultimately how I ended up fixing all of my selector issues. By being careful with classnames and `dataset` contents, I was able to solve ~90% of my contextmenu item selector issues. 

The trickiest cases all involved situations in which I needed to distinguish between two different areas, but for which there were no mutually exclusive selectors. For these cases, I either rearranged my existing DOM, or I added an [extra invisible `div`](https://github.com/jupyterlab/jupyterlab/pull/5409/files#r222194968) to help distinguish one area from another.
